custom-metrics-stackdriver-adapter/pkg/adapter/provider/cache.go

```diff
@@ -0,0 +1,77 @@
+/*
+Copyright 2025 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+		http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package provider
+
+import (
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/metrics/pkg/apis/external_metrics"
+	"sigs.k8s.io/custom-metrics-apiserver/pkg/provider"
+)
+
+// Cache for GetExternalMetric responses.
+type externalMetricCache struct {
```

Comment:
Should we add a cache size?  The cache size is growing with more HPA requests and might be OOM.

Reply:
Ah good point.. We may want to eventually do some garbage collection, otherwise the cache may grow really big.

For most clusters with a fairly static set of HPAs, the number of entries should stay fairly stable but in clusters with a lot of churn, I could imagine this getting very big.

So perhaps we need a separate thread that is periodically cleaning up the cache?

Reply:
For v1, I think I'd prefer as simple of a cache as possible. The OOM problem can be solved with giving the addon-resizer extra memory. I haven't verified but I believe the cache entries here should be really small, even with hundreds of HPAs, the memory footprint should be quite small.

Reply:
The LRUExpireCache looks great here, I think we just need to be careful with it because it isn't threadsafe but looks like it can be an almost straight drop-in replacement for the cache here :) Good find!